retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java

```diff
@@ -0,0 +1,190 @@
+ *                              method not available</li>
+ *     <li>{@code httpStatus} - {@link HttpStatus#code()}</li>
+ * </ul>
+ */
+final class RetrofitClassAwareMeterIdPrefixFunction extends RetrofitMeterIdPrefixFunction {
+
+    private static final List<Class<?>> RETROFIT_ANNOTATIONS = ImmutableList.of(
+            POST.class, PUT.class, PATCH.class, HEAD.class, GET.class, OPTIONS.class, HTTP.class, DELETE.class
+    );
+    private static final String METHOD_TAG_NAME = "method";
+    private static final String PATH_TAG_NAME = "path";
+    private static final String HTTP_METHOD_TAG_NAME = "httpMethod";
+    private static final List<Tag> DEFAULT_METHOD_TAGS = ImmutableList.of(
+            Tag.of(METHOD_TAG_NAME, "UNKNOWN"),
+            Tag.of(PATH_TAG_NAME, "UNKNOWN")
+    );
+
+    private final Map<String, List<Tag>> methodNameToTags;
+    private final String name;
+    private final String serviceName;
+    private final String serviceTagName;
+
+    /**
+     * Returns a newly created {@link RetrofitClassAwareMeterIdPrefixFunction} with the specified {@code name}
+     * and {@code serviceClass}.
+     */
+    public static RetrofitMeterIdPrefixFunction of(String name, Class<?> serviceClass) {
+        return builder(name, serviceClass).build();
+    }
+
+    /**
+     * Returns a newly created {@link RetrofitMeterIdPrefixFunctionBuilder} with the specified {@code name}
+     * and {@code serviceClass}.
+     */
+    public static RetrofitMeterIdPrefixFunctionBuilder builder(String name, Class<?> serviceClass) {
+        return new RetrofitMeterIdPrefixFunctionBuilder(requireNonNull(name, "name"),
+                                                        requireNonNull(serviceClass, "serviceClass"));
+    }
+
+    RetrofitClassAwareMeterIdPrefixFunction(String name,
+                                            @Nullable String serviceTagName,
+                                            Class<?> serviceClass) {
+        super(name, null, null);
+
+        this.name = name;
+        this.serviceTagName = firstNonNull(serviceTagName, "service");
+        this.serviceName = serviceClass.getSimpleName();
+        this.methodNameToTags = defineTagsForMethods(serviceClass);
+    }
+
+    @Override
+    public MeterIdPrefix activeRequestPrefix(MeterRegistry registry, RequestLog log) {
+        final ImmutableList.Builder<Tag> tagsListBuilder = ImmutableList.builder();
+        buildTags(tagsListBuilder, log);
+        return new MeterIdPrefix(name, tagsListBuilder.build());
+    }
+
+    @Override
+    public MeterIdPrefix apply(MeterRegistry registry, RequestLog log) {
+        final ImmutableList.Builder<Tag> tagListBuilder = ImmutableList.builder();
+        buildTags(tagListBuilder, log);
+        RequestMetricSupport.appendHttpStatusTag(tagListBuilder, log);
+        return new MeterIdPrefix(name, tagListBuilder.build());
+    }
+
+    private void buildTags(ImmutableList.Builder<Tag> tagsBuilder, RequestLog log) {
```

Comment:
How about changing `RetrofitMeterIdPrefixFunction.buildTags` to `protected`? 
And remove the overrided methods:
- `activeRequestPrefix(MeterRegistry, RequestLog)`
- `apply(MeterRegistry, RequestLog)`?

Reply:
Usage of overriding methods gives us ability for a bit of microoptimization by providing expected list size. That by claims from guava may work faster then unbounded immutable list builder.

Reply:
I guess it's fine as it is unless @ikhoon has a strong opinion on this.

Reply:
> Usage of overriding methods gives us ability for a bit of microoptimization by providing expected list size. That by claims from guava may work faster then unbounded immutable list builder.

Sorry I missed this message. 

> I guess it's fine as it is unless @ikhoon has a strong opinion on this.

I don't have strong opinion here. Let's leave it as it is. :-)

Reply:
Oops, I found out that we don't use exact number for the expected size [here](https://github.com/line/armeria/blob/master/retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitMeterIdPrefixFunction.java#L89).
Could you change that as well please?
On my second thought, perhaps we can set the `builderWithExpectedSize` with the number + 1 in `RetrofitMeterIdPrefixFunction`, so that we can remove `activeRequestPrefix()` and `apply()` in `RetrofitClassAwareMeterIdPrefixFunction` and just overriding `buildTags()`.
Either way is fine for me though. 😉